src/establish-metamask-repository.ts

```diff
@@ -0,0 +1,259 @@
+  cachedRepositoriesDirectoryPath,
+  outputLogger,
+}: {
+  repositoryReference: string;
+  workingDirectoryPath: string;
+  cachedRepositoriesDirectoryPath: string;
+  outputLogger: AbstractOutputLogger;
+}): Promise<ExistingRepository> {
+  const {
+    repositoryShortname,
+    repositoryDirectoryPath,
+    repositoryDirectoryExists,
+    isKnownMetaMaskRepository,
+  } = await resolveRepositoryReference({
+    repositoryReference,
+    workingDirectoryPath,
+    cachedRepositoriesDirectoryPath,
+  });
+  const isGitRepository = await directoryExists(
+    path.join(repositoryDirectoryPath, '.git'),
+  );
+
+  if (repositoryDirectoryExists && !isGitRepository) {
+    throw new Error(
+      `"${repositoryDirectoryPath}" is not a Git repository, cannot proceed.`,
+    );
+  }
+
+  let branchInfo: BranchInfo;
+  if (isGitRepository) {
+    log('Repository has been cloned already to', repositoryDirectoryPath);
+    branchInfo = await getBranchInfo(repositoryDirectoryPath);
+  } else {
+    branchInfo = await cloneRepository({
+      repositoryShortname,
+      repositoryDirectoryPath,
+      cachedRepositoriesDirectoryPath,
+      outputLogger,
+    });
+  }
+
+  return {
+    ...branchInfo,
+    shortname: repositoryShortname,
+    directoryPath: repositoryDirectoryPath,
+    isKnownMetaMaskRepository,
+  };
+}
+
+/**
+ * Clones a MetaMask repository using the `gh` tool.
+ *
+ * @param args - The arguments to this function.
+ * @param args.repositoryShortname - The name of the repository minus the
+ * organization (so, "utils" instead of "MetaMask/utils").
+ * @param args.repositoryDirectoryPath - The path where the repository should be
+ * cloned to.
+ * @param args.cachedRepositoriesDirectoryPath - The parent directory in which
+ * to keep the new repository.
+ * @param args.outputLogger - Writable streams for output messages.
+ * @throws If `repositoryDirectoryPath` is not within
+ * `cachedRepositoriesDirectoryPath`, or you do not have `gh` installed.
+ */
+async function cloneRepository({
+  repositoryShortname,
+  repositoryDirectoryPath,
+  cachedRepositoriesDirectoryPath,
+  outputLogger,
+}: {
+  repositoryShortname: string;
+  repositoryDirectoryPath: string;
+  cachedRepositoriesDirectoryPath: string;
+  outputLogger: AbstractOutputLogger;
+}) {
+  /* istanbul ignore next: There's no way to reproduce this; this is just to be absolutely sure */
+  if (
+    path.dirname(repositoryDirectoryPath) !== cachedRepositoriesDirectoryPath
+  ) {
+    throw new Error(
+      'You seem to be pointing to a directory outside the cached repositories directory. Refusing to proceed to avoid data loss.',
+    );
+  }
+
+  log('Assuming', repositoryShortname, 'is the name of a MetaMask repo');
+
+  log('Removing existing', repositoryDirectoryPath);
+  await fs.promises.rm(repositoryDirectoryPath, {
+    recursive: true,
+    force: true,
+  });
+
+  log('Cloning', repositoryShortname, 'to', repositoryDirectoryPath);
+  outputLogger.logToStdout(
```

Comment:
This seems to be the only place this output logger is used so far. I'm curious why this was developed rather than using `console.log` here directly, or some other logging library. The benefits aren't totally clear at this point

Reply:
There will be other places where this logger is used in future commits. Regardless, I decided to create this pattern for a couple of reasons:

1. I am already using this pattern for input. That is, `process.argv` is passed into `main` (in `main.ts`) rather than having `main` read `process.argv` directly. This allows `process.argv` to be faked easily. So it seemed logical to have main take `process.stdout` and `process.stderr` as well.

3. It offers a way to specify and see the expected output in tests more easily. The executable prints out multiple lines of output. And if you want to output multiple lines to the screen, you can either choose to divide them up into multiple `console.log` calls, or put them together into one call. If we spy on `console.log` or `console.error`, it would take a bit of trial and error to figure out how many calls were being made. I feel like this is too low of a level to think about from an executable perspective. You shouldn't have to care how exactly messages end up getting added to an output stream. You shouldn't even have to really care about the stream! All we really care about is, what text does stdout or stderr hold? The output logger interface allows you to say this.